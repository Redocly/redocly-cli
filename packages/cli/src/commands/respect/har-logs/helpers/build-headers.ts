/**
 * Support the three possible header formats we'd get from a request or
 * response:
 *
 * - A flat array with both names and values: [name, value, name, value, ...]
 * - An object with array values: { name: [value, value] }
 * - An object with string values: { name: value }
 */
export function buildHeaders(headers: any): any[] {
  const list = [];
  if (Array.isArray(headers)) {
    for (let i = 0; i < headers.length; i += 2) {
      list.push({
        name: headers[i],
        value: headers[i + 1],
      });
    }
  } else if (headers instanceof Map || headers.entries) {
    // Handle both Map and Headers objects (which have entries())
    for (const [name, value] of headers.entries()) {
      list.push({ name, value });
    }
  } else if (typeof headers === 'object') {
    for (const [name, value] of Object.entries(headers)) {
      list.push({ name, value: value as string });
    }
  }
  return list;
}
