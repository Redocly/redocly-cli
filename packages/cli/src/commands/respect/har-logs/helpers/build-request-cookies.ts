import * as cookie from 'cookie';

export function buildRequestCookies(headers: any): any[] {
  const cookies: any[] = [];
  for (const [header, headerValue] of Object.entries(headers)) {
    if (header.toLowerCase() === 'cookie') {
      // Handle both array and single string cases
      const cookieValues = Array.isArray(headerValue) ? headerValue : [headerValue];
      for (const value of cookieValues) {
        const parsed = cookie.parse(value);
        for (const [name, value] of Object.entries(parsed)) {
          cookies.push({ name, value });
        }
      }
    }
  }
  return cookies;
}
