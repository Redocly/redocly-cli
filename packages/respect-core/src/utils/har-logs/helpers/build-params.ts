import * as querystring from 'querystring';

export function buildParams(paramString: string): any[] {
  const params = [];
  const parsed = querystring.parse(paramString);
  for (const name in parsed) {
    const value = parsed[name];
    if (Array.isArray(value)) {
      value.forEach((item) => {
        params.push({ name, value: item });
      });
    } else {
      params.push({ name, value });
    }
  }
  return params;
}
