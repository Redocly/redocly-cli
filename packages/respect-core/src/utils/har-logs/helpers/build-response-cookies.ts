import * as setCookie from 'set-cookie-parser';

export function buildResponseCookies(headers: any): any[] {
  const cookies: any[] = [];
  // Handle both plain objects and Headers instance
  const setCookies =
    headers instanceof Headers
      ? headers.getSetCookie() // For Headers object
      : headers['set-cookie']; // For plain objects

  if (setCookies) {
    for (const headerValue of setCookies) {
      let parsed;
      try {
        parsed = setCookie.parse(headerValue);
      } catch {
        continue;
      }
      for (const cookie of parsed) {
        const { name, value, path, domain, expires, httpOnly, secure } = cookie;
        const harCookie: any = {
          name,
          value,
          httpOnly: httpOnly || false,
          secure: secure || false,
        };
        if (path) {
          harCookie.path = path as string | undefined;
        }
        if (domain) {
          harCookie.domain = domain as string | undefined;
        }
        if (expires) {
          harCookie.expires = expires.toISOString();
        }
        cookies.push(harCookie);
      }
    }
  }
  return cookies;
}
