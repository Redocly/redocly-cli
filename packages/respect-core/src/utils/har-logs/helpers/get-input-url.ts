export function getInputUrl(input: any): URL {
  const url = typeof input === 'string' ? input : input.url;
  return new URL(url);
}
