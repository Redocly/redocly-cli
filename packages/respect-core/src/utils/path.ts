import { isBrowser } from '@redocly/openapi-core';

let pathLib: any;

if (isBrowser) {
  // Browser environment
  pathLib = await import('path-browserify');
} else {
  // Node.js environment
  pathLib = await import('node:path');
}

export default pathLib;
