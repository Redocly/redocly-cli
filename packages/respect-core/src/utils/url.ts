export function combineUrl(host: string, path: string): string {
  const normalizedHost = host.endsWith('/') ? host.slice(0, -1) : host;
  const normalizedPath = path.startsWith('/') ? path.slice(1) : path;
  return `${normalizedHost}/${normalizedPath}`;
}
