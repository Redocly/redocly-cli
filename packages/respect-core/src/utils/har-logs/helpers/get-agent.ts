import * as http from 'http';
import * as https from 'https';
import { Dispatcher } from 'undici';
import { getInputUrl } from './get-input-url';
import { handleRequest } from './handle-request';

const HarHttpAgent = createAgentClass(http.Agent);
const HarHttpsAgent = createAgentClass(https.Agent);

let globalHttpAgent: any;
let globalHttpsAgent: any;

// Add new Undici dispatcher
class HarDispatcher extends Dispatcher {
  constructor(opts?: any) {
    super(opts);
  }

  dispatch(options: Dispatcher.DispatchOptions, handler: Dispatcher.DispatchHandlers): boolean {
    // Handle HAR logging here similar to handleRequest
    handleRequest({ input: options, handler, harLog: new Map(), isUndici: true });
    return super.dispatch(options, handler);
  }
}

export function getAgent(input: any, options: any): any {
  // Add Undici dispatcher support
  if (options.dispatcher) {
    if (options.dispatcher instanceof Dispatcher) {
      return new HarDispatcher(options.dispatcher);
    }
    return options.dispatcher;
  }

  if (options.agent) {
    if (typeof options.agent === 'function') {
      return function (...args: any[]) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const agent = options.agent.call(this, ...args);
        if (agent) {
          instrumentAgentInstance(agent);
          return agent;
        }
        return getGlobalAgent(input);
      };
    }
    instrumentAgentInstance(options.agent);
    return options.agent;
  }
  return getGlobalAgent(input);
}

function getGlobalAgent(input: any): any {
  const url = getInputUrl(input);
  if (url.protocol === 'http:') {
    if (!globalHttpAgent) {
      globalHttpAgent = new HarHttpAgent();
    }
    return globalHttpAgent;
  }
  if (!globalHttpsAgent) {
    globalHttpsAgent = new HarHttpsAgent();
  }
  return globalHttpsAgent;
}

/**
 * Instrument an existing Agent instance. This overrides the instance's
 * `addRequest` method. It should be fine to continue using for requests made
 * without `withHar` - if the request doesn't have our `x-har-request-id`
 * header, it won't do anything extra.
 */
function instrumentAgentInstance(agent: any): void {
  const { addRequest: originalAddRequest } = agent;
  if (!originalAddRequest.isHarEnabled) {
    agent.addRequest = function addRequest(request: any, ...args: []) {
      handleRequest(request, ...args);
      return originalAddRequest.call(this, request, ...args);
    };
    agent.addRequest.isHarEnabled = true;
  }
}

function createAgentClass(BaseAgent: any): any {
  class HarAgent extends BaseAgent {
    constructor(...args: any[]) {
      super(...args);
      (this.addRequest as any).isHarEnabled = true;
    }

    addRequest(request: any, ...args: []): void {
      handleRequest(request, ...args);
      return super.addRequest(request, ...args);
    }
  }

  return HarAgent;
}
