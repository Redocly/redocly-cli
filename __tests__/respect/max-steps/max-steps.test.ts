import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { getCommandOutput, getParams } from '../../helpers.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

test('should quit an infinite loop on RESPECT_MAX_STEPS', () => {
  const indexEntryPoint = join(process.cwd(), 'packages/cli/lib/index.js');
  const fixturesPath = join(__dirname, 'arazzo.yaml');
  const args = getParams(indexEntryPoint, [
    'respect',
    fixturesPath,
    '--verbose',
    '--workflow',
    'infinite',
  ]);

  const result = getCommandOutput(args, {
    RESPECT_MAX_STEPS: '10',
  });

  expect(result).toMatchSnapshot();
}, 120_000);
