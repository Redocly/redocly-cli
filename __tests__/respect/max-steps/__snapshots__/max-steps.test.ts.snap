// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should quit an infinite loop on RESPECT_MAX_STEPS 1`] = `
"────────────────────────────────────────────────────────────────────────────────

  Running workflow arazzo.yaml / infinite

  ✗ GET /ping - step pre-step

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action continue for the step pre-step

  Running child workflow for the step with-nested
  Running workflow arazzo.yaml / nested-workflow

  ✗ GET /ping - step step-1

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action next-step for the step step-1

  ✗ GET /ping - step step-2

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request



  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping

  ✗ GET /ping - step ping

    Request URL: https://bad-api-url.com/api/ping

    ✗ failed network request

  Running failure action infinite-loop for the step ping










  Failed tests info:

  Workflow name: infinite

    stepId - pre-step
    ✗ failed network request
      fetch failed
      
    stepId - step-1
    ✗ failed network request
      fetch failed
      
    stepId - step-2
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ failed network request
      fetch failed
      
    stepId - ping
    ✗ maximum steps reached
    Reason: Max steps (10) reached
  Summary for arazzo.yaml
  
  Workflows: 1 failed, 1 total
  Steps: 11 failed, 11 total
  Checks: 11 failed, 11 total
  Time: <test>ms


┌─────────────────────────────────────────────────────┬────────────┬─────────┬─────────┬──────────┐
│ Filename                                            │ Workflows  │ Passed  │ Failed  │ Warnings │
├─────────────────────────────────────────────────────┼────────────┼─────────┼─────────┼──────────┤
│ x arazzo.yaml                                       │ 1          │ 0       │ 1       │ -        │
└─────────────────────────────────────────────────────┴────────────┴─────────┴─────────┴──────────┘


 Tests exited with error 

"
`;
